nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,772 @@
+            unsafe {
+                if buf.is_empty() {
+                    epin[i].ptr.write(|w| w.bits(0));
+                } else {
+                    epin[i].ptr.write(|w| w.bits(ram_buf.as_ptr() as u32));
+                }
+                epin[i].maxcnt.write(|w| w.maxcnt().bits(buf.len() as u8));
+            }
+
+            if i == 0 {
+                // EPIN0: a short packet (len < max_packet_size0) indicates the end of the data
+                // stage and must be followed by us responding with an ACK token to an OUT token
+                // sent from the host (AKA the status stage) -- `usb-device` provides no call back
+                // for that so we'll trigger the status stage using a shortcut
+                let is_short_packet = buf.len() < self.max_packet_size_0 as usize;
+                regs.shorts.modify(|_, w| {
+                    if is_short_packet {
+                        w.ep0datadone_ep0status().set_bit()
+                    } else {
+                        w.ep0datadone_ep0status().clear_bit()
+                    }
+                });
+
+                // Hack: send status stage if the IN transfer is not acknowledged after a few frames
+                let frame_counter = regs.framecntr.read().framecntr().bits();
+                let ep0_state = self.ep0_state.borrow(cs);
+                let mut state = ep0_state.get();
+                state.in_transfer_state = TransferState::Started(frame_counter);
+                ep0_state.set(state);
+            }
+
+            // Clear ENDEPIN[i] flag
+            regs.events_endepin[i].reset();
+
+            // Kick off device -> host transmission. This starts DMA, so a compiler fence is needed.
+            dma_start();
+            regs.tasks_startepin[i].write(|w| w.tasks_startepin().set_bit());
+            while regs.events_endepin[i].read().events_endepin().bit_is_clear() {}
+            regs.events_endepin[i].reset();
+            dma_end();
+
+            // Clear EPSTATUS.EPIN[i] flag
+            regs.epstatus.write(|w| unsafe { w.bits(1 << i) });
+
+            Ok(buf.len())
+        })
+    }
+
+    fn read(&self, ep_addr: EndpointAddress, buf: &mut [u8]) -> usb_device::Result<usize> {
+        if !self.is_used(ep_addr) {
+            return Err(UsbError::InvalidEndpoint);
+        }
+
+        if ep_addr.is_in() {
+            return Err(UsbError::InvalidEndpoint);
+        }
+
+        let i = ep_addr.index();
+        interrupt::free(|cs| {
+            let regs = self.periph.borrow(cs);
+
+            // Control EP 0 is special
+            if i == 0 {
+                // Control setup packet is special, since it is put in registers, not a buffer.
+                if regs.events_ep0setup.read().events_ep0setup().bit_is_set() {
+                    regs.events_ep0setup.reset();
+
+                    let ep0_state = unsafe { &mut *self.ep0_state.borrow(cs).as_ptr() };
```

Comment:
can you avoid the unsafe bits?  They look a bit magical/noisy:

```suggestion
                    let mut ep0_state = self.ep0_state.borrow(cs);
```

then just pass `&mut ep0_state` below?

Alternatively, pass the `cs: &CriticalSection` to `read_control_setup` and borrow `ep0_state` in there?